### ISSUE TITLE
@SkipSelf() cd: ChangeDetectorRef doesn't return parent ChangeDetector

### ISSUE THREAD

**Issue by johncrim** [angular/angular#39475]
# 🐞 bug report

### Affected Package
@angular/core/di

### Is this a regression?
I'm not sure, but given [this workaround](https://github.com/angular/angular/issues/22560#issuecomment-473958144) for bug #22560 it probably worked pre-Ivy.

### Description

It's not possible for a component to obtain the `ChangeDetectorRef` or `Injector` from its parent by using @SkipSelf. eg:

```ts
  constructor(
    // Using parent ChangeDetector, b/c host bindings are part of the parent CD.
    // Not intuitive, but true: https://github.com/angular/angular/issues/22560
    @SkipSelf() changeDetector: ChangeDetectorRef
  ) {
```

There is no compile error or runtime error, but the argument passed in is the component's ChangeDetectorRef or Injector.

The bug is apparent in `getOrCreateInjectable()`:
```ts
export function getOrCreateInjectable<T>(
    tNode: TDirectiveHostNode|null, lView: LView, token: Type<T>|InjectionToken<T>,
    flags: InjectFlags = InjectFlags.Default, notFoundValue?: any): T|null {
  if (tNode !== null) {
    const bloomHash = bloomHashBitOrFactory(token);
    // If the ID stored here is a function, this is a special object like ElementRef or TemplateRef
    // so just call the factory function to create it.
    if (typeof bloomHash === 'function') {
      enterDI(lView, tNode);
      try {
        const value = bloomHash();
        if (value == null && !(flags & InjectFlags.Optional)) {
          throw new Error(`No provider for ${stringifyForError(token)}!`);
        } else {
          return value;
        }
      } finally {
        leaveDI();
      }
    } else if (typeof bloomHash == 'number') {
      if (bloomHash === -1) {
        // `-1` is a special value used to identify `Injector` types.
        return new NodeInjector(tNode, lView) as any;
      }
```

If the injection token is one of (ChangeDetectorRef, Injector, ElementRef, ViewContainerRef, TemplateRef) it has a `NG_ELEMENT_ID` that is a function, which results in all `InjectFlags` being ignored except `InjectFlags.Optional`.

In combination with this bug ( [InjectFlags ignored by NodeInjector.get()](https://github.com/angular/angular/issues/31776#issuecomment-717398208) ) and this by-design bug ( [HostBindings not updated by detectChanges() in OnPush component](https://github.com/angular/angular/issues/22560) ), there's no reasonable way to update HostBindings in OnPush components.

I think it's reasonable that some of the unsupported injection tokens don't support `@SkipSelf()` (eg TemplateRef and ViewContainerRef make sense; Injector is debatable), but due to HostBindings only being updated by the parent ChangeDetector, we need a straightforward way to obtain the parent ChangeDetector. Also, if `@SkipSelf()` (or other flags) aren't supported, an error should be raised (compile time or runtime) instead of injecting the wrong argument.

## 🔬 Minimal Reproduction
This is a karma test running in StackBlitz
https://stackblitz.com/edit/ng10-bug-skipself

Test source:
```ts
import { ChangeDetectorRef, Component, Injectable, Injector, SkipSelf } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';

@Injectable()
class TestService {
  constructor(public id: string) { }
}

@Component({
  selector: 'app-child',
  template: `<div>nothing to see here</div>`,
  providers: [
    {
      provide: TestService,
      useFactory: () => new TestService('child')
    }
  ]
})
class ChildComponent {

  constructor(
    public changeDetector: ChangeDetectorRef,
    @SkipSelf() public parentChangeDetector: ChangeDetectorRef,
    public injector: Injector,
    @SkipSelf() public parentInjector: Injector,
    @SkipSelf() public testService: TestService

  ) { }

}

@Component({
  selector: 'app-parent',
  template: `<app-child></app-child>`,
  providers: [
    {
      provide: TestService,
      useFactory: () => new TestService('parent')
    }
  ]
})
class ParentComponent {

  constructor(
    public changeDetector: ChangeDetectorRef,
    public injector: Injector
  ) { }

}

describe('Component injection with @SkipSelf', () => {

  let fixture: ComponentFixture<ParentComponent>;
  let child: ChildComponent;

  beforeEach((() => {
    TestBed.configureTestingModule({
      declarations: [
        ChildComponent,
        ParentComponent
      ],
    }).compileComponents();

    fixture = TestBed.createComponent(ParentComponent);
    child = fixture.debugElement.children[0].componentInstance as ChildComponent;
  }));

  it('child can get parent Injector (FAILS)', () => {
    expect((child.parentInjector as any)._lView)
      .toBe((fixture.componentInstance.injector as any)._lView);
  });

  it('child can get parent ChangeDetector (FAILS)', () => {
    expect((child.parentChangeDetector as any)._cdRefInjectingView)
      .toBe((fixture.componentInstance.changeDetector as any)._cdRefInjectingView);
  });

  it('inject(@SkipSelf Injector) is different from inject(Injector) (FAILS)', () => {
    expect((child.parentInjector as any)._lView)
      .not.toBe((child.injector as any)._lView);
  });

  it('inject(@SkipSelf Injector) returns service from parent (FAILS)', () => {
    const parentInjectorService = child.parentInjector.get(TestService);
    const childInjectorService = child.injector.get(TestService);
    expect(parentInjectorService.id)
      .not.toBe(childInjectorService.id);
  });

  it('inject(@SkipSelf ChangeDetectorRef) is different from inject(ChangeDetectorRef) (FAILS)', () => {
    expect((child.parentChangeDetector as any)._cdRefInjectingView)
      .not.toBe((child.changeDetector as any)._cdRefInjectingView);
  });

  it('inject(@SkipSelf TestService) returns service from parent component (PASSES)', () => {
    expect(child.testService.id)
      .toBe('parent');
  });

});
```

## 🌍  Your Environment

**Angular Version:**
<pre><code>
Angular CLI: 10.0.8
Node: 14.8.0
OS: win32 x64

Angular: 10.0.14
... animations, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router
Ivy Workspace: Yes

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.1000.8
@angular-devkit/build-angular     0.1000.8
@angular-devkit/build-optimizer   0.1000.8
@angular-devkit/build-webpack     0.1000.8
@angular-devkit/core              10.0.8
@angular-devkit/schematics        10.0.8
@angular/cli                      10.0.8
@ngtools/webpack                  10.0.8
@schematics/angular               10.0.8
@schematics/update                0.1000.8
rxjs                              6.5.5
typescript                        3.9.7
webpack                           4.43.0
</code></pre>

**Comment by JoostK** [comment]
Thanks for putting this together! It's been reported in #34066 and #34819, which is being addressed in #39464. As such, I'll close this as duplicate.

**Comment by johncrim** [comment]
Thanks! My fault for not finding those before creating the repro cases.
